kubespawner/reflector.py

```diff
@@ -140,20 +173,34 @@ def _watch_and_update(self):
                 if not self.first_load_future.done():
                     # signal that we've loaded our initial data
                     self.first_load_future.set_result(None)
+
+                # in case of timeout_seconds, the w.stream just exits (no exception thrown)
+                # -> we stop the watcher and start a new one
+
                 for ev in w.stream(
                         getattr(self.api, self.list_method_name),
-                        self.namespace,
-                        label_selector=self.label_selector,
-                        resource_version=resource_version,
+                        **watch_args
                 ):
                     cur_delay = 0.1
                     resource = ev['object']
+                    if not resource.metadata.resource_version:
+                        # reset resources_version if we don't have one
+                        del watch_args['resource_version']
+                    elif (watch_args.get('resource_version') != resource.metadata.resource_version):
+                        # only set resources_version if new resource_version is not None
+                        watch_args['resource_version'] = resource.metadata.resource_version
```

Comment:
I don't think this resource_version update is accomplishing its goal. We are starting the list fresh with `._list_and_update` on every restart, and must use that resource_version in order to avoid lost events. Recording the resource_version here could allow us to skip that step on subsequent reconnects, but we would need to be very careful to ensure that we *don't* call `_list_and_update` on resume. Instead, it's probably simpler to drop this resource_version tracking and start fresh with `_list_and_update` each go. This will restart rarely, so the millisecond-level cost of a fresh check is probably worth the simpler logic.

Reply:
I completely skipped the `._list_and_update`. An initial watch without resource version retrieves the full list. After reading the Kubernetes API docs again, I found that it has become more specific about this. Watch only returns the full list since the beginning of history, which may not return the full current list. I'll bring that back in the cleanup